### DISCUSSION TITLE
🛠 Improvements to OA v3 to achieve W3C interoperability

### DISCUSSION THREAD

**Discussion by HJunyuan** [Open-Attestation/open-attestation#249] [General]
## Context

There have been some concerns (e.g. #233, #234, etc.) that the Verifiable Credentials (VCs) generated by the OpenAttestation framework may not be interoperable with other W3C VC wallets/platforms in the wild.

To resolve these concerns, we have established 3 action items laid out in this document:

1. Make sure OA produces a VC data model that aligns with the [W3C Verifiable Credentials Data Model](https://www.w3.org/TR/vc-data-model/) specifications
2. Make sure OA passes the https://github.com/w3c-ccg/vc-api-test-suite
3. Publish the proof format of `OpenAttestationMerkleProofSignature2018` onto W3C and IETF

## Objective

This document aims to document proposed changes and facilitate collaboration from the community to achieve W3C interoperability.

To facilitate ease of communication, we will use the following alias:

- `OA v3`: The current state of OpenAttestation v3
- `OA v4`: The proposed changes laid out in this document

### Action 1a: Omit the use of ~~`version`~~ and leave `credentialSchema` empty

<details>
<summary>In OA v3, a wrapped (or signed/issued) VC may look something like this:</summary>

```jsonc
{
  "version": "https://schema.openattestation.com/3.0/schema.json",
  "issuer": {
    // ...
  },
  "credentialSubject": {
    // ...
  },
  "proof": {
     // ...
   }
  // ...
}
```
</details>

However, the `version` field that is currently used in OA v3 is not part of the [VC base context](https://www.w3.org/TR/vc-data-model/#base-context). For the OA framework to easily identify the version of the VC, OA v4 will rely on the expanded URI where `@type` contains `https://schema.openattestation.com/4.0/credentials#OpenAttestationCredential`.

Example of a OA v4 VC in an expanded document form:
```jsonc
[
  {
    "@type": [
      "https://www.w3.org/2018/credentials#VerifiableCredential",
      "https://schema.openattestation.com/4.0/credentials#OpenAttestationCredential"
    ],
    // ...
  }
]
```

**Action 1a Conclusion**:

- OA v4 will no longer utilise the ~~`version`~~ field
- OA v4 will rely on the expanded URI to determine the OA version of a VC
- Leave the [`credentialSchema`](https://www.w3.org/TR/vc-data-model/#data-schemas) field empty for issuers to insert at their discretion to describe the schema of `credentialSubject`

### Action 1b: Consolidate all issuer-related information into `issuer` field

<details>
<summary>In OA v3, a DID-signed VC may look something like this:</summary>

```jsonc
{
  "issuer": {
    "id": "https://example.com",
    "type": "OpenAttestationIssuer",
    "name": "DEMO STORE"
  },
  "credentialSubject": {
    // ...
  },
  "openAttestationMetadata": {
    "template": {
      // ...
    },
    "proof": {
      "type": "OpenAttestationProofMethod",
      "method": "DID",
      "value": "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733",
      "revocation": {
        "type": "NONE"
      }
    },
    "identityProof": {
      "type": "DNS-DID",
      "identifier": "demo-tradetrust.openattestation.com"
    }
  },
  "proof": {
    "type": "OpenAttestationMerkleProofSignature2018",
    "proofPurpose": "assertionMethod",
    "targetHash": "d876067507c616c42a5c32a42bc37754403533246c04445678f024d73f18b745",
    "proofs": [
      "239cf91faee989c730299cc1683b9211a36547107625d26a06f8ad5323d5e406",
      "bed332534ffb2cfcc40a5c2c71141dbda5c21f114085f783a437cc0f2d5a86c3"
    ],
    "merkleRoot": "dc34b7bc4e707c77327db76536625c81c2a6777934df566b261bd8ec3f24f8c4",
    "salts": "W3sidmFsdWU...",
    "privacy": {
      "obfuscated": []
    },
    "key": "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733#controller",
    "signature": "0x9fc61bfdf30553d7e2ecd77fbd190c44bb2f3192760ebc08eb32fa3d2c15626b3af3577c0abcfa0877e656380be9eceb4e2a665e5ad9324eaeddb7d90ae3b5311c"
  }
}
```
</details>

The current verification process in OA v3 only verifies against the [`proof.key` field](https://github.com/Open-Attestation/oa-verify/blob/a1f892f70d8997632035b79ada97f6b30e672a57/src/verifiers/documentStatus/didSigned/didSignedDocumentStatus.ts#L224) and [`openAttestationMetadata.proof.value` field](https://github.com/Open-Attestation/oa-verify/blob/a1f892f70d8997632035b79ada97f6b30e672a57/src/verifiers/documentStatus/didSigned/didSignedDocumentStatus.ts#L225). The current OA v3 verification process **does not involve** any part of the `issuer` field which may cause irregularities with external verifiers.

<details>
<summary>In contrast, the W3C data model specifications states that the <a href="https://www.w3.org/TR/vc-data-model/#issuer" target=”_blank”><code>issuer</code> field</a> should be utilised:</summary>

```jsonc
// Example 8:
{
  "issuer": "https://example.edu/issuers/14",
  // ...
}

// OR

// Example 9:
{
  "issuer": {
    "id": "did:example:76e12ec712ebc6f1c221ebfeb1f",
    "name": "Example University"
  },
  // ...
}
```
</details>

In OA v4, issuance will be described in their respective fields:
- `issuer` - Issuer and identity related information
- `credentialStatus` - Revocation status (if any)
- `renderMethod` - Rendering instructions (if any)

```jsonc
{
  "@context": [
    "https://www.w3.org/ns/credentials/v2",
    "https://schemata.openattestation.com/com/openattestation/4.0/alpha-context.json"
  ],
  "name": "Republic of Singapore Driving Licence",
  "type": ["VerifiableCredential", "OpenAttestationCredential"],
  "issuer": {
    "id": "did:ethr:0xB26B4941941C51a4885E5B7D3A1B861E54405f90",
    "type": "OpenAttestationIssuer",
    "name": "Government Technology Agency of Singapore (GovTech)",
    "identityProof": {
      "identityProofType": "DNS-DID",
      "identifier": "example.openattestation.com"
    }
  },
  "validFrom": "2021-03-08T12:00:00+08:00",
  "credentialSubject": {
    "id": "urn:uuid:a013fb9d-bb03-4056-b696-05575eceaf42",
    "type": ["DriversLicense"],
    "name": "John Doe",
    "licenses": [
      {
        "class": "3",
        "description": "Motor cars with unladen weight <= 3000kg",
        "effectiveDate": "2013-05-16T00:00:00+08:00"
      },
      {
        "class": "3A",
        "description": "Motor cars with unladen weight <= 3000kg",
        "effectiveDate": "2013-05-16T00:00:00+08:00"
      }
    ]
  },
  "credentialStatus": {
    "id": "https://ocsp-sandbox.openattestation.com",
    "type": "OpenAttestationOcspResponder"
  },
  "proof": {
    "type": "OpenAttestationMerkleProofSignature2018",
    "proofPurpose": "assertionMethod",
    "targetHash": "1074606ce1e0e44434b214336e02d7e5b7fc035b1fc3f527ae4eb14f7023a02e",
    "proofs": [],
    "merkleRoot": "1074606ce1e0e44434b214336e02d7e5b7fc035b1fc3f527ae4eb14f7023a02e",
    "salts": "W3sidmFsdWUiOiI2YWI2ODNlMDBhMDQ3OTU0ODBhNjE4ZjhjMWY3NDFjODZhYjFiZTAxN2U0ZWU2Mjk4NmZlY2FkNDIzMjFkZWIzIiwicGF0aCI6IkBjb250ZXh0WzBdIn0seyJ2YWx1ZSI6ImFkM2U3N2E5OGFhYTFlNjhiMzQyZWE3YzAyN2E3MjA0Nzg2OTY1MzkxZTE5MjFmNWUwM2Y1Y2NmMTE4N2U5OGEiLCJwYXRoIjoiQGNvbnRleHRbMV0ifSx7InZhbHVlIjoiNzdkMTg3MjdlY2Q3ODJlNDg4NjQ0ZWQ3NDU3ZTZlM2VlMmRjYWUxY2Q4NzBhZjg5YTkxMWNhYTIwMWIyYzVjZSIsInBhdGgiOiJuYW1lIn0seyJ2YWx1ZSI6IjFkYTVkMjM1ZGZhMzhjYjhmZmE4NTczNGY3YmJiMjRlYmY1YzE3ODhkZGNlNjQ2ZjhjMDU2MGIxYTU5ZGIwM2IiLCJwYXRoIjoidHlwZVswXSJ9LHsidmFsdWUiOiI4NWZhYzMxYzg3YzIxYzVjMjU4NzcxZmFmN2I5NmVjMjZkODk4OGU1MjkzZTg4MTYxODQzM2JlNDlmOGZhMjg3IiwicGF0aCI6InR5cGVbMV0ifSx7InZhbHVlIjoiNTIyODQ0YWNlZDdmNDFhY2M1NjFiZmRhNjU3YzI3ZTkzODI0NzhhYTEyODdhMmEzMmI3ZTg3NjU5ZDg5Njg0NiIsInBhdGgiOiJpc3N1ZXIuaWQifSx7InZhbHVlIjoiODZmODQwMTY3MTE5NzZiNGQyYTQ5ZDhjMzY0NTEyMjNhMTYwNWY2Njc0OGQ3NWFjOTEwN2I4OGNkMDNmMGU4OCIsInBhdGgiOiJpc3N1ZXIudHlwZSJ9LHsidmFsdWUiOiIzNDhiZGQzNGM2YWJkYjQ4YmM5ZTE0YWE5NDJlODg5ZTY0MTlhMTM5NDk4NTQ5Zjg1NjBiMzc1ZjVjODA1YTY1IiwicGF0aCI6Imlzc3Vlci5uYW1lIn0seyJ2YWx1ZSI6IjVlOWQ2NWM2Nzg5NWZkNjMwODRhNTI2MDhmOTNkOTA2YjgyZDllYzI2N2Q0NTAwNmU2N2E1NTdlMTllMGQ0MDciLCJwYXRoIjoiaXNzdWVyLmlkZW50aXR5UHJvb2YuaWRlbnRpdHlQcm9vZlR5cGUifSx7InZhbHVlIjoiNGY3MGJhZGQxODk1NGM3ZmJjNTY4MjNmMTdlYjA5YTdkYjUyNzkyYTk5ZmVkYTVhNzM2NDg0MmExZjNjODU2OSIsInBhdGgiOiJpc3N1ZXIuaWRlbnRpdHlQcm9vZi5pZGVudGlmaWVyIn0seyJ2YWx1ZSI6IjczODBhN2FhYTUxYTdmNGE3MzQ4MDVlODYzZGY3NTI2NmU4NzU5NzQ1Mzc4MGFiMDU5MWRmNzE3M2I4NmYyYjQiLCJwYXRoIjoidmFsaWRGcm9tIn0seyJ2YWx1ZSI6IjVlZGIzODAxYjY3MGVjNGMwNmNmYjJlNmY0NTFmZDQzM2ZkNjYyMTA1M2Q1ZjljZWZlNDJkNDMzYTc3ZTc0MDciLCJwYXRoIjoiY3JlZGVudGlhbFN1YmplY3QuaWQifSx7InZhbHVlIjoiZWEyNDAxYzAyOGUyMzZiM2M0MGM0MDI4N2VlYTg2MWJiMGE0MDA1NTAwYzY1OGY0ZDZlN2Q3NzdmMmZkZDRkMiIsInBhdGgiOiJjcmVkZW50aWFsU3ViamVjdC50eXBlWzBdIn0seyJ2YWx1ZSI6IjNmMjYwYThiYTlkMGZlYTkzZGQ1NGM2ZDc4YzIzZDZlOWUxNmU1MmYwOGM2OTYzMTJhMmE5ZWU3NmVlMjY1YjEiLCJwYXRoIjoiY3JlZGVudGlhbFN1YmplY3QubmFtZSJ9LHsidmFsdWUiOiJkMDcwMWIyNzUxNjAxM2JhNzUzYThjMzhmNjg0MGE4ZjEyYjI3OGQyNDMwZGU1MzUyODg5ODRkYTA1ODhmZjUwIiwicGF0aCI6ImNyZWRlbnRpYWxTdWJqZWN0LmxpY2Vuc2VzWzBdLmNsYXNzIn0seyJ2YWx1ZSI6IjMwZjdmNTc4M2ZkYmI4YjBmMDVjY2MzMjQ1ZWYyYTU3NjRiYzMxNDlkYzcwZGNhMzgwOTc2NzJlOWE5NTA0ZjMiLCJwYXRoIjoiY3JlZGVudGlhbFN1YmplY3QubGljZW5zZXNbMF0uZGVzY3JpcHRpb24ifSx7InZhbHVlIjoiNzFiODE5MzA2YmU5OWU4YjBmODUxMWMwYzhiMWU1MWZlOWQwYmNjYmQ2YjJjNTg2ZjdjMTZiNDc5ZDEwOWVmZSIsInBhdGgiOiJjcmVkZW50aWFsU3ViamVjdC5saWNlbnNlc1swXS5lZmZlY3RpdmVEYXRlIn0seyJ2YWx1ZSI6IjliNTM2ZjE4OWM2YzVmZmU0NDkwMWIwNjA0ODkzOWUyZGM2YTRkMTk4Njc3Zjk1NjA3M2VkOTQ1ZTQyZDM2YTMiLCJwYXRoIjoiY3JlZGVudGlhbFN1YmplY3QubGljZW5zZXNbMV0uY2xhc3MifSx7InZhbHVlIjoiZDgxODMzNDMzMjVlYzRkOTU3YzgyZDBjMTk1ZWE3MzM4NTM2YWI5ZDg4NjZkODk3NjBlZGQxNDE0MmZmNzc3ZSIsInBhdGgiOiJjcmVkZW50aWFsU3ViamVjdC5saWNlbnNlc1sxXS5kZXNjcmlwdGlvbiJ9LHsidmFsdWUiOiJlOWViY2FmMTM0ZjcwZjdlZThjNWU2ODVjYTgwNjMwOWNjZWY4MWVlMGJhNmJlYzRhYTJjOTBlNDkzMzhjNzVlIiwicGF0aCI6ImNyZWRlbnRpYWxTdWJqZWN0LmxpY2Vuc2VzWzFdLmVmZmVjdGl2ZURhdGUifSx7InZhbHVlIjoiYjZlZGFkNzRiZWVkZjRiNzYwNjA3YzkwMTY3YmJmY2JkZmY3YmRjNGM4MWY3ZjYzMjU0NTEzODBhZDg0NjQyMyIsInBhdGgiOiJjcmVkZW50aWFsU3RhdHVzLmlkIn0seyJ2YWx1ZSI6ImNlMjRkOTM5MWY3ODJhMmY2N2RlM2U2ZTAyMGVmMDA5MGRiOGMzNWI5YzhiMGIwZWJiZWI2M2Y5ZWNjOTM3ZTIiLCJwYXRoIjoiY3JlZGVudGlhbFN0YXR1cy50eXBlIn0seyJ2YWx1ZSI6IjIxNTg3NmUzNmFlY2EzZDU2YTVmN2FkODhkNzdmYmIwZTI2ODBmOTU5NjU4M2MzMzFmOTY0YjIzNTBjNTczYjciLCJwYXRoIjoicHJvb2YudHlwZSJ9LHsidmFsdWUiOiJkMDVhNGI3ZTdmMGNhYjlmYjUyNzlmODk0M2E5NmUyYWQ1YTIzYzRkMzc5MWQ5ZGViMDY1OGE5MzdjY2RlZmMzIiwicGF0aCI6InByb29mLnByb29mUHVycG9zZSJ9LHsidmFsdWUiOiI2MjI0MmQzODU0ZGI5ZTc0NGQ2ODU3ZjAyM2E2YzlhNDFmNDNhOWFlYTU0ODJmMmE4YjQ3ZmQ0MDJhNGVhMDAwIiwicGF0aCI6InByb29mLnRhcmdldEhhc2gifSx7InZhbHVlIjoiMTMxZjJiOGUyM2VhYjAzOTg3NDdjZmQ3MDJlNWZjNDc0NzZiM2E5MDcyMzgwNDUzY2U1NTU5NDFmN2NhMWJiYSIsInBhdGgiOiJwcm9vZi5wcm9vZnMifSx7InZhbHVlIjoiOWM4OTQ1NDAyMjI0YWM2Y2U4ODcyMDg0OTgwYjBmNDMwODZiM2U2N2QzYzM3NjhiMGUzZjRkNWNmNWRkZjg1MyIsInBhdGgiOiJwcm9vZi5tZXJrbGVSb290In0seyJ2YWx1ZSI6IjI1Yjc5ZGFiYzVkY2I5ODQ2Nzc5OGQ2MGIxOGVlZDhhZjI2NjI5MGQzZjJhMWJhNTAxMDViNDJkMTk5MzE3MGUiLCJwYXRoIjoicHJvb2Yuc2FsdHMifSx7InZhbHVlIjoiNzJiOGFkNDdiODFkODgwMzNlZmYzYmYyMzIzYWYwY2EzNjgyYjA3NWE5NTMxNjgxYzQ0ZjdiYTA1MTA4MGYyNSIsInBhdGgiOiJwcm9vZi5wcml2YWN5Lm9iZnVzY2F0ZWQifSx7InZhbHVlIjoiYzU1YjdjZmNlYTZlMjdmZmMxMGMzNWUxMjhjNmQ0YmMyMzE5M2Y4Y2NjNDg5OGUxOWZiYzRhNDdmOTNiOWNjNCIsInBhdGgiOiJwcm9vZi5rZXkifSx7InZhbHVlIjoiNGMxZTZmODY0YTJiZDQ1OTQ3YjJlOWZmMzlkYzM5OGM1NWQ1NTE1ZGJjMmY3OWMwMGU3ZjcyZjk0N2I0YmU1YiIsInBhdGgiOiJwcm9vZi5zaWduYXR1cmUifSx7InZhbHVlIjoiODRjMjE5MDA5MDA4NDVkNzgzNDU1OTI2ZGY0MWM2MDgxMTE5OGNmNTIwNWU3ZDQ5Y2YyODk0ZDgwNDQyZjk4MCIsInBhdGgiOiJyZW5kZXJNZXRob2RbMF0uaWQifSx7InZhbHVlIjoiOWFhMjE4OTFiYjY0NDM0OGY3OTcyNDc3M2UxYzNmZTJmMWNiMjE0YzgyYWQ5Y2QyZjM3Y2RjN2Q1OTgyNDhlNiIsInBhdGgiOiJyZW5kZXJNZXRob2RbMF0udHlwZSJ9LHsidmFsdWUiOiJiZmQxY2Q2NjEwODdjN2Q0NDNhZjY2MzhiYWEwMWU4NWM4NDhhNzQwNDYwYjA2OWY2NDU0NDJhMWE4ZmYwZWI2IiwicGF0aCI6InJlbmRlck1ldGhvZFswXS50ZW1wbGF0ZU5hbWUifV0=",
    "privacy": { "obfuscated": [] },
    "key": "did:ethr:0xB26B4941941C51a4885E5B7D3A1B861E54405f90#controller",
    "signature": "0xbb59f3df32d6eb53bccc8a17fb2ced77bfd01b68168e57bdfdc05b7d1e7eda5048cc964b19deb90dc788b2efa9b6c80fd5be068c74848d7ca534336c4ca1fd061c"
  },
  "renderMethod": [
    {
      "id": "https://demo-renderer.opencerts.io",
      "type": "OpenAttestationEmbeddedRenderer",
      "templateName": "GOVTECH_DEMO"
    }
  ]
}
```

> Refer to Action 1e for the corresponding JSON-LD context definition

**Action 1b Conclusion**:

- The verification process of OA v4 will now only utilise the `issuer.id` field instead of the `openAttestationMetadata.proof.value` field.
- OA v4 will no longer utilise the ~~`openAttestationMetadata`~~ field
- OA v4 will put issuance information into the respective fields: `issuer`, `credentialStatus` and `renderMethod`

### Action 1c: Supporting other DID methods (e.g. `did:web:xxx`, etc.)

In the current oa-verify implementation:
- The [default resolver already includes the web-did-resolver](https://github.com/Open-Attestation/oa-verify/blob/e562ce9fe9e1a9b2742cb40e6476bb0661abbda0/src/did/resolver.ts#L32) to resolve DIDs like `did:web:example.com`
- However, [only the `EcdsaSecp256k1RecoveryMethod2020` signature method is supported](https://github.com/Open-Attestation/oa-verify/blob/e562ce9fe9e1a9b2742cb40e6476bb0661abbda0/src/did/verifier.ts#L99-L105)
- As such, if one were to have their `did:web:xxx` resolved to the same DID Document of their existing `did:ethr:xxx`, verification may continue to work (though some code changes may be required)

<details>
<summary>For example, here's the DID Document of <code>did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733</code>:</summary>

```json
{
  "@context": [
    "https://www.w3.org/ns/did/v1",
    "https://w3id.org/security/suites/secp256k1recovery-2020/v2"
  ],
  "id": "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733",
  "verificationMethod": [
    {
      "id": "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733#controller",
      "type": "EcdsaSecp256k1RecoveryMethod2020",
      "controller": "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733",
      "blockchainAccountId": "eip155:1:0x1245e5B64D785b25057f7438F715f4aA5D965733"
    }
  ],
  "authentication": [
    "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733#controller"
  ],
  "assertionMethod": [
    "did:ethr:0x1245e5B64D785b25057f7438F715f4aA5D965733#controller"
  ]
}
```
</details>

> One could [create](https://w3c-ccg.github.io/did-method-web/#create-register) `did:web:w3c-ccg.github.io:user:alice` which points to `https://w3c-ccg.github.io/user/alice/did.json` and supply the same DID Document content as above

**Action 1c Conclusion**
- [Optional / TBC] Decide on a [list of signature methods](https://w3c-ccg.github.io/security-vocab/) to be supported in OA v4 (other than `EcdsaSecp256k1RecoveryMethod2020`)
- [Optional / TBC] With the support of those signature methods, how would end-users generate the key pairs? (i.e. need to find the corresponding key generation libraries like `ethers.js` for end-users to utilise)

### Action 1d: `issuer.identityProof` will be redundant if other DID methods

With Action 1c implemented, the `DNS-DID` identity proof that OA currently uses (i.e. creating a DNS TXT record to declare ownership of the domain) is no longer necessary if a `did:web:xxx` is used for issuance since the nature of resolving this DID is already establishing a link with the domain.

**Action 1d Conclusion**
- In OA v4, `identityProof` should be optional if a `did:web:xxx` (or any other similar domain-backed DID) is used

### Action 1e: Reduce/consolidate OpenAttestation specific @contexts

In OA v4, a single JSON-LD context definition is used to give semantic meaning to all OpenAttestation specific fields and can be declared in this way:

```jsonc
{
  "@context": [
    "https://www.w3.org/ns/credentials/v2",
    "https://schemata.openattestation.com/com/openattestation/4.0/alpha-context.json"
  ],
  "type": ["VerifiableCredential", "OpenAttestationCredential"],
  // ...
}
```

The proposed OA v4 JSON-LD context definition (<code>https://schemata.openattestation.com/com/openattestation/4.0/alpha-context.json</code>):

> Try it out in JSON-LD Playground: https://tinyurl.com/2ba8jl5e

```json
{
  "@context": {
    "@version": 1.1,
    "@protected": true,
    "schema": "https://schema.org/",
    "oacred": "https://schemata.openattestation.com/com/openattestation/4.0/alpha-context.json#",
    "OpenAttestationCredential": {
      "@id": "oacred:OpenAttestationCredential",
      "@context": {
        "name": "schema:name"
      }
    },
    "OpenAttestationIssuer": {
      "@id": "oacred:OpenAttestationIssuer",
      "@context": {
        "name": "schema:name",
        "identityProof": {
          "@id": "oacred:identityProof",
          "@context": {
            "identityProofType": "oacred:identityProofType",
            "identifier": {
              "@id": "oacred:identityProofIdentifier",
              "@type": "schema:identifier"
            }
          }
        }
      }
    },
    "OpenAttestationOcspResponder": {
      "@id": "oacred:credentialStatus"
    },
    "OpenAttestationEmbeddedRenderer": {
      "@id": "oacred:renderMethod",
      "@context": {
        "templateName": "oacred:renderTemplateName"
      }
    },
    "OpenAttestationMerkleProofSignature2018": {
      "@id": "oacred:OpenAttestationMerkleProofSignature2018",
      "@context": {
        "proofPurpose": "oacred:proofPurpose",
        "targetHash": "oacred:proofTargetHash",
        "proofs": {
          "@id": "oacred:proofIntermediateProofs",
          "@container": "@list"
        },
        "merkleRoot": "oacred:proofMerkleRoot",
        "salts": "oacred:proofSalts",
        "privacy": {
          "@id": "oacred:proofPrivacy",
          "@context": {
            "obfuscated": {
              "@id": "oacred:proofPrivacyObfuscated",
              "@container": "@list"
            }
          }
        },
        "key": "oacred:proofKey",
        "signature": "oacred:proofSignature"
      }
    }
  }
}
```

**Action 1e Conclusion**

- A minimum of 2 contexts are required: `VerifiableCredential` and `OpenAttestationCredential`
- The context for issuer-specific fields (e.g. `credentialSubject`) will need to be defined/provided by each issuer

## Action 2: Test against the VC API test suite

VC API Test Suite: https://github.com/w3c/vc-data-model-2.0-test-suite
Findings:
- 17 May 2023 - https://gist.github.com/cavacado/28d1c50746b8149e22ceac04759eea51

## Action 3: Publishing OAMerkleProofSignature2018 specification

W3C draft ([source](https://github.com/Open-Attestation/OAMerkleProofSignature2018)): https://open-attestation.github.io/OAMerkleProofSignature2018/

IETF draft: ⚠️ [Work in progress]

**Comment by cavacado** [comment]
Nice work just writing up and summarising the actions to be taken.
General QN:
why not just major version the schema and make v4 so as to have a clean break from all the other legacy docs?
To me, once you touch the OpenAttestationContexts, you'll probably be moving alot of things to warrant a major upgrade.

For action 1b:
Would it be a trivial change to oa-verify to use the `issuer.id` field instead of `document.openAttestationMetadata.identityProof.identifier` in this [line](https://github.com/Open-Attestation/oa-verify/blob/master/src/verifiers/issuerIdentity/did/didIdentityProof.ts#L121)? since i think did-signed oa docs do put the did:eth in there (maybe its only me who does it lol)

For action 1c:
Currently we are picking and choosing the did-resolvers that we currently support; would it be better to just depend a dockerised universal resolver of our own 
https://github.com/decentralized-identity/universal-resolver

For action 1d:
Yea i dont know how to handle the legacy documents if we nuke this section (maybe im too noob)

**Reply by HJunyuan** [reply]
**In response to your general qn**:
I would agree with the v4 approach and have a clean break from all the other legacy docs. I believe the original intention was to provide backwards compatibility (esp how we are unsure of the adoption of OA v3) so I prefaced this document with v3.1. If all is okay, we can move towards v4 cleanly without "excess baggage"

**For Action 1b**:
I originally thought it would be a trivial change to just a single `issuer.id`. However, I realised a few things:

- For DID issuance, the [`verifySignature()`](https://github.com/Open-Attestation/oa-verify/blob/a1f892f70d8997632035b79ada97f6b30e672a57/src/did/verifier.ts#L78-L84) method requires the DID and key (e.g. `did:ethr:xxx` and `did:ethr:xxx#controller` respectively)
- For document store issuance, the content of `issuer.id` may not be so straight forward. Should the document store address live in this field? Based on what I understand from the W3C VC spec, there are [external VS embedded proofs](https://www.w3.org/TR/vc-data-model/#proofs-signatures) that the document store address can live

Correct me if I'm wrong, but I think the line you referenced to is only for `DID` type of issuer identity check? More specifically, OA issuer identity has 3 types:
1. `DID` (no domain linkage, just DID alone)
2. `DNS-DID` (with domain linkage for DID-issued VCs)
3. `DNS-TXT` (with domain linkage for document store issued VCs)

As such, `did:eth:0x123xxx` is placed in `openAttestationMetadata.identityProof.identifier` instead of a domain

**For Action 1c**:
Yes, we should make use of universal resolver for the respective types of DIDs. In fact, the default resolver that comes with `oa-verify` [already includes the `did:web:xxx` resolver](https://github.com/Open-Attestation/oa-verify/blob/a1f892f70d8997632035b79ada97f6b30e672a57/src/did/resolver.ts#L32)

**For action 1d**:
Perhaps implementing as v4 would make it easier for us to achieve this

**Comment by simboonlong** [comment]
some thoughts on backward compatibility on decentralised renderer side https://github.com/Open-Attestation/tutorial-renderer/pull/1

**Reply by HJunyuan** [reply]
Yes, the respective decentralised renderers from each issuer would need to be updated to extract the new fields since it is essentially a schema change

+1 to your PR and the latter solution to create an updated v3 example with a correct schema that utilises the fields in `credentialSubject`. Will need to make sure it matches a corresponding `@context` as well I think

**Comment by arpentnoir** [comment]
**Action 1a: Utilise credentialSchema instead of version**

I'm not sure this fits with the purpose of the credentialSchema - the primary aim there is to allow issuers to include a schema defining the syntax of the claims represented in the credentialSubject. I think there are two distinct processes being merged here:

1. allowing a verifier to check conformance of the credentialSubject object against some published schema (e.g. the UniversityDegreeCredential example schema given here https://www.w3.org/TR/vc-data-model/#data-schemas)
2. allowing the oa-verify library to determine the expected structure of the credential as a whole so as to correctly verify the signature. 

for the purposes of action 1a, I think this could easily be achieved by just adding the type `OpenAttestationCredential` and having this defined in the oa v4 context. the oa-verify library could just use the expanded URI to determine the OA version. 

**Action 1b: Consolidate all issuer-related information into issuer field**

there are two objects defined in the example OAv4 issuer object example above that map to concepts defined in the VC data model and are expected to be keys in the root of the credential.

OA "revocation" -> VCDM "credentialStatus" (see https://www.w3.org/TR/vc-data-model/#status)
OA "renderer" -> VCDM "renderMethod" (see https://www.w3.org/TR/vc-data-model-2.0/#reserved-extension-points)

the latter is of course just a reserved word in the draft v2 spec, but you can see where they are headed...

I think rather than including "revocation" and "renderer" nested under the "issuer" object, the approach more in line with the VCDM would be to publish both these mechanisms as credential status and render method specs and use the standard structure to reference them.

**Reply by HJunyuan** [reply]
Hi @arpentnoir, thank you for your inputs on this. To make sure we understand you correctly:

**Action 1a: Utilise `credentialSchema` instead of version**
Thanks for pointing out the clash in our interpretation of this field.

- We will leave the `credentialSchema` field empty for issuers to insert at their discretion
- For the oa-verify library to easily determine the OA version of a VC, we will rely on the expanded URI indicating OA v4 (`https://schema.openattestation.com/4.0/credentials#OpenAttestationCredential`):
  ```jsonc
  {
    "@type": [
      "https://www.w3.org/2018/credentials#VerifiableCredential",
      "https://schema.openattestation.com/4.0/credentials#OpenAttestationCredential"
    ]
    // ...
  }
  ```

**Action 1b: Consolidate all issuer-related information into issuer field**
Agreed, the `credentialStatus` and `renderMethod` fields do seem to be more appropriate for this case.

<details>
  <summary>Here's a snippet of what it could look like:</summary>

  ```jsonc
  {
    // ...
    "type": ["VerifiableCredential", "OpenAttestationCredential"],
    "issuer": {
      "id": "did:ethr:0xE39479928Cc4EfFE50774488780B9f616bd4B830",
      "type": "OpenAttestationIssuer",
      "name": "Government Technology Agency of Singapore (GovTech)",
      "identityProof": {
        "type": "DNS-DID",
        "identifier": "donotverify.testing.verify.gov.sg"
      }
    },
    "credentialStatus": {
      "@type": "OpenAttestationCredentialStatus",
      "type": "OCSP_RESPONDER",
      "location": "https://ocsp-responder.example.com"
    },
    "renderMethod": {
      "@type": "OpenAttestationRenderMethod",
      "type": "EMBEDDED_RENDERER",
      "name": "GOVTECH_DEMO",
      "url": "https://demo-renderer.opencerts.io"
    },
    "credentialSubject": {
      "id": "urn:uuid:a013fb9d-bb03-4056-b696-05575eceaf42",
      "type": ["DriversLicense"],
      "name": "John Doe",
      "licenses": [
        {
          "class": "3",
          "description": "Motor cars with unladen weight <= 3000kg",
          "effectiveDate": "2013-05-16T00:00:00+08:00"
        },
        {
          "class": "3A",
          "description": "Motor cars with unladen weight <= 3000kg",
          "effectiveDate": "2013-05-16T00:00:00+08:00"
        }
      ]
    }
  }
  ```
</details>

Also understand that the `renderMethod` field is at risk of being removed from the specification. My worry is that it gets removed after OA v4 has already adopted it. Do you think it will pose any problems down the line if that were to occur?

> Implementers MAY use these properties, but SHOULD expect them and/or their meanings to change during the process to normatively specify them. Implementers SHOULD NOT use these properties without a publicly disclosed specification describing their implementation.

**Comment by HJunyuan** [comment]
For those following this thread, [OpenAttestation v4 (beta) has been released](https://github.com/Open-Attestation/open-attestation/releases/tag/v6.10.0-beta.1) to incorporate the changes listed here (i.e. aligning with the W3C VC Data Model)

**Comment by HJunyuan** [comment]
Closing this thread in favour of #309 to consolidate any last feedback/suggestions before we finalise the release of OpenAttestation OA v4.0
